### ISSUE TITLE
Non blocking bio_connect moves to BIO_CONN_S_OK incorrectly

### ISSUE THREAD

**Issue by nikolaydio** [openssl/openssl#8034]
#### Edit

> It addition to BIO_should_read() and BIO_should_write() it is also possible for BIO_should_io_special() to be true during the initial connection process with the reason BIO_RR_CONNECT. If this is returned then this is an indication that a connection attempt would block, the application should then take appropriate action to wait until the underlying socket has connected and retry the call.

It seems like it is expected from the application code to handle the case "-1 return, BIO_should_io_special == true, BIO_get_retry_reason == BIO_RR_CONNECT? It seems strange to require going down in the abstraction layers to figure out if we can continue polling. I really would expect for BIO_do_connect(and BIO_read/BIO_write) to continue returning -1 + BIO_should_retry until it actually connects....

---------------------
I'm currently trying to setup non blocking sockets in windows.
Please, note that the exact same code behaves and works exactly as I would expect in linux and the issue is Windows only. 

The issue seems to be in the following piece of code:

https://github.com/openssl/openssl/blob/09abbca13f985d57e872ea070dcc53b283460d52/crypto/bio/bss_conn.c#L152-L163

I also used the following link for reference:
https://docs.microsoft.com/en-us/windows/desktop/api/winsock2/nf-winsock2-connect

### Reproduction steps(roughly)
1. Make a `BIO_do_connect(bio);` call  (returns -1, should_retry = true):
1.1 (See the above code) BIO_connect call happens. It will answer "hey we failed, but you should try later"
1.2 Transition to the following state `c->state = BIO_CONN_S_BLOCKED_CONNECT;`
2. Make a `BIO_do_connect(bio);` call  (returns 1)
https://github.com/openssl/openssl/blob/09abbca13f985d57e872ea070dcc53b283460d52/crypto/bio/bss_conn.c#L185-L198
2.1 No error will be reported, as of course there were no socket calls at all, and the BIO will transition to BIO_CONN_S_OK and the user will receive "1".
 #### In other words the bio reports that the connection has been established, but it is in fact not ready yet.
3. Make BIO_read/BIO_write call. Receive -1 and retry "false". The last winsock error is WSAENOTCONN.
The application code should interpret this as "we got disconnected or something bad happened", however this is not true at all. 
Placing a sleep after the connect to give it some time resolves the issue as it gives the socket some time to establish the connection and so the following read/write calls are accepted.

**Comment by nikolaydio** [comment]
I believe the BIO should be calling
```
     ret = BIO_connect(b->num, BIO_ADDRINFO_address(c->addr_iter),  BIO_SOCK_KEEPALIVE | c->connect_mode); 
```
until it gets confirmation that the connection is established, and then and only then it should transition to BIO_CONN_S_OK.

See microsoft's documentation

> Until the connection attempt completes on a nonblocking socket, all subsequent calls to connect on the same socket will fail with the error code WSAEALREADY, and WSAEISCONN when the connection completes successfully

**Comment by tvdijk** [comment]
I'm seeing the same problem, also on Windows. I'm not explicitly calling `BIO_do_connect()`, just `BIO_write()` (I'm using a `BIO_new_ssl_connect` chain, which will connect under the hood as needed). The second call to `BIO_write()` returns -1, `BIO_should_retry()` returns `false`, and the last WinSock error is `WSANOTCONN`. Simply retrying allows everything to proceed normally. I'm on OpenSSL 1.0.2q.

**Comment by t8m** [comment]
As this was reported against 1.1.1 and 1.0.2 versions, please open a new issue if this still affects your use-case with currently supported OpenSSL versions.
